### PR TITLE
refactor(package-rules): move to class based implementation

### PR DESCRIPTION
**Pull request by secustor**
## Changes
This PR refactors the package-rules to a class based approach based on the datasource setup. 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/package-rules/current-version/index.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        }
+      } else {
+        logger.debug(
+          { matchCurrentVersionStr, currentValue },
+          'Could not find a version to compare'
+        );
+        return false;
+      }
+    }
+    return positiveMatch;
```
Comment (nabeelsaabna) [inline lib/util/package-rules/current-version/index.ts:83]:
```suggestion
    const unconstrainedValue = !!lockedVersion && is.undefined(currentValue);
    const version = allVersioning.get(versioning);
    const matchCurrentVersionStr = matchCurrentVersion.toString();
    const matchCurrentVersionPred = configRegexPredicate(
      matchCurrentVersionStr
    );

    if (matchCurrentVersionPred) {
      if (
        !unconstrainedValue &&
        (!currentValue || !matchCurrentVersionPred(currentValue))
      ) {
        return false;
      }
      return true;
    } 

	if (version.isVersion(matchCurrentVersionStr)) {
      let isMatch = false;
      try {
        isMatch =
          unconstrainedValue ||
          !!(
            currentValue &&
            version.matches(matchCurrentVersionStr, currentValue)
          );
      } catch (err) {
        // Do nothing
      }
      if (!isMatch) {
        return false;
      }
      return true;
    } 
	
	
    const compareVersion =
      currentValue && version.isVersion(currentValue)
	    ? currentValue // it's a version so we can match against it
	    : lockedVersion ?? currentVersion; // need to match against this currentVersion, if available

    if (compareVersion) {
	  // istanbul ignore next
	  if (version.isVersion(compareVersion)) {
	    const isMatch = version.matches(compareVersion, matchCurrentVersion);
	    // istanbul ignore if
	    if (!isMatch) {
	  	  return false;
	    }
	    return true;
	  } 
	  return false;
	}

	logger.debug(
	  { matchCurrentVersionStr, currentValue },
	  'Could not find a version to compare'
	);
	return false;
```

Comment (nabeelsaabna) [reply lib/util/package-rules/current-version/index.ts:83]: tried to make sure it's correct according to the [code-simplicity](https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#code-simplicity) new guide lines

Comment (nabeelsaabna) [reply lib/util/package-rules/current-version/index.ts:83]: might be a good idea to ignore my suggestion and do it in the IDE to make sure all curly braces are handled correctly

Comment (secustor) [reply lib/util/package-rules/current-version/index.ts:83]: This is simply copied so far from the original file and I haven't made pass to improve it. Thanks for the suggestion.

Comment (nabeelsaabna) [reply lib/util/package-rules/current-version/index.ts]: I can be done in a asperate PR later as well 👍

File: lib/util/package-rules/index.spec.ts [hunk 2]
```diff
@@ -167,9 +167,9 @@ describe('util/package-rules', () => {
       ],
     };
     const res = applyPackageRules(dep);
-    expect(res.automerge).toBeTrue();
-    const res2 = applyPackageRules({ ...dep, depName: 'anything' });
-    expect(res2.automerge).toBeFalse();
+    expect(res.automerge).toBeFalse();
+    const res2 = applyPackageRules({ ...dep, depName: 'foo' });
+    expect(res2.automerge).toBeTrue();
```
Comment (viceice) [inline lib/util/package-rules/index.spec.ts:172]: why is this test changed? we shouldn't need to change any test in this complex refactor.

**Review by viceice — CHANGES_REQUESTED** [review]
Do we really need al those subdirectories? i would prefer to rename all `package-rules/<dir>/index.ts` to `package-rules/<dir>.ts` if there is only one file (excluding `.spec.ts` files)

**Comment by viceice** [comment]
otherwise i like the refactoring, should make the code much more maintainable

File: lib/util/package-rules/index.spec.ts (hunk 2 above)
Comment (secustor) [reply lib/util/package-rules/index.spec.ts:172]: I think this is a bug/false test, based on the conversation on slack with @rarkins 

If no match is applied it should default to match all. In the original test case that would mean both results would be false as the package rule would apply to both.

**Comment by secustor** [comment]
> Do we really need al those subdirectories? i would prefer to rename all `package-rules/<dir>/index.ts` to `package-rules/<dir>.ts` if there is only one file (excluding `.spec.ts` files)

No there is no reason in this PR, my guess has been it would be easier to automate doc creation later on if we have folders much like for datasources. That way we can easier add additional files to it such as README files.

File: lib/util/package-rules/index.spec.ts (hunk 2 above)
Comment (viceice) [reply lib/util/package-rules/index.spec.ts:172]: We should extract this fix to a separate PR, just to be sure

**Comment by viceice** [comment]
> > Do we really need al those subdirectories? i would prefer to rename all `package-rules/<dir>/index.ts` to `package-rules/<dir>.ts` if there is only one file (excluding `.spec.ts` files)
> 
> No there is no reason in this PR, my guess has been it would be easier to automate doc creation later on if we have folders much like for datasources. That way we can easier add additional files to it such as README files.

I don't think we will generate seperate docs for those parts

File: lib/util/package-rules/api.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import { PackageNameMatcher } from './package-names';
+import { PackagePatternsMatcher } from './package-patterns';
+import { PackagePrefixesMatcher } from './package-prefixes';
+import { PathsMatcher } from './paths';
+import { SourceUrlPrefixesMatcher } from './sourceurl-prefixes';
+import { SourceUrlsMatcher } from './sourceurls';
+import type { MatcherApi } from './types';
+import { UpdateTypesMatcher } from './update-types';
+
+const api = new Map<string, MatcherApi[]>();
```
Comment (viceice) [inline lib/util/package-rules/api.ts:17]: i don't see any usage for the keys. shouldn't we use a simple array here?
```suggestion
const api MatcherApi[][] = [];
```

File: lib/util/package-rules/base-branches.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import is from '@sindresorhus/is';
+import type { PackageRule, PackageRuleInputConfig } from '../../config/types';
+import { configRegexPredicate } from '../regex';
+import { Matcher } from './base';
+
+export class BaseBranchesMatcher extends Matcher {
+  static readonly id: string = 'base-branches';
```
Comment (viceice) [inline lib/util/package-rules/base-branches.ts:7]: i don't see a value for those id's, ad they are only used in the api map, but then we only use the `map.values`

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/package-rules/api.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import { PackageNameMatcher } from './package-names';
+import { PackagePatternsMatcher } from './package-patterns';
+import { PackagePrefixesMatcher } from './package-prefixes';
+import { PathsMatcher } from './paths';
+import { SourceUrlPrefixesMatcher } from './sourceurl-prefixes';
+import { SourceUrlsMatcher } from './sourceurls';
+import type { MatcherApi } from './types';
+import { UpdateTypesMatcher } from './update-types';
+
+const api: MatcherApi[][] = [];
```
Comment (viceice) [inline lib/util/package-rules/api.ts:17]: we should name this `matchers`, `api` seems to unspecific here.

File: lib/util/package-rules/package-patterns.spec.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      const result = packageNameMatcher.matches(
+        {
+          depName: undefined,
+        },
+        {
+          matchPackagePatterns: ['@opentelemetry/http'],
+        }
+      );
+      expect(result).not.toBeNull();
+      expect(result).toBeFalse();
```
Comment (viceice) [inline lib/util/package-rules/package-patterns.spec.ts:17]:
```suggestion
      expect(result).toBeFalse();
```
i think this should be enough

File: lib/util/package-rules/index.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+export function applyPackageRules<T extends PackageRuleInputConfig>(
+  inputConfig: T
+): T {
+  let config = { ...inputConfig };
+  const packageRules = config.packageRules ?? [];
+  logger.trace(
+    { dependency: config.depName, packageRules },
+    `Checking against ${packageRules.length} packageRules`
+  );
+  packageRules.forEach((packageRule) => {
```
Comment (viceice) [inline lib/util/package-rules/index.ts:77]: use `for...of` loop

**Review by viceice — APPROVED** [review]

**Pull request merged**